### ISSUE TITLE
[tune] Implementing HyperOpt Save and Restore

### ISSUE THREAD

**Issue by tawe141** [ray-project/ray#5150]
Does Tune have a warm start feature? Is it dependent on the algorithm I'm using?

**Comment by richardliaw** [comment]
Certain training algorithms have a warm start, such as PBT. Are you looking for anything in particular/can you tell me more about your use case?

**Comment by tawe141** [comment]
I'm currently running hyperparameter tuning with HyperOptSearch. Does this
have warm starting?

**Comment by richardliaw** [comment]
Ah, in those cases, right now you'll have to implement it yourself. One option would be to do something like:

```
class Trainable():
   ...
   def _setup(self, config):
       if "restore_from" in config:
           self.restore(config["restore_from"])
       ...

trainable = Trainable()
for i in range(...):
    trainable.train()

save_path = trainable.save()
tune.run(Trainable, config={"restore_from": save_path}
```

Would that work for you?

**Comment by tawe141** [comment]
Sorry, I should clarify. I want to warm-start the hyperparameter tuning itself, not the neural networks it's training.

**Comment by richardliaw** [comment]
@tawe141, sorry - what exactly do you mean by "warm-start"? Does it mean that you want to warmup the system? Or does it mean you want to provide a set of pre-defined points to "seed" the optimization? Or something else?

**Comment by tawe141** [comment]
The latter. I would like to use a past run I did, extract the hyperparameters and losses from those, and have the suggestion algorithm suggest the next set of hyperparameters. I know I can set `points_to_evaluate` in HyperOptSearch, but I don't want to re-evaluate this set of parameters when I've already tried those previously.

**Comment by richardliaw** [comment]
OK I see; it's possible to do this through HyperOpt APIs: https://github.com/hyperopt/hyperopt/issues/267. You can access the hyperOpt trial object with `HyperOptSearch._hpopt_trials()`. 

It's out of band, but it should work. If you end up getting it working, I'd be more than happy to help you get it merged into master! And also if you run into issues, feel free to ask.

**Comment by jredondopizarro** [comment]
Hi @richardliaw, related with this issue, I have tried to set `HyperOptSearch._hpopt_trials() `to be a trials object from a past run, as you suggested. However, it seems like these manually added trials are not being taken into account by the suggestion algorithm. 

For example: 
 - First experiment: I run `HyperOptSearch() `for 25 iterations, save its trials object, feed this trials object into a new `HyperOptSearch()` instance and run this new instance for another 25 iterations. 
- Second experiment: I run `HyperOptSearch()` for 50 iterations at once.

Of course, the hyperparameter space, the function to optimize and the random state are the same in both cases. Then, I would expect to get the same results from both experiments, but it turns out that the last 25 iterations from the first experiment are different from the last 25 iterations from the second experiment.

Do you have any clue about  this issue?
Thanks.

**Comment by richardliaw** [comment]
Hey,

Can you post your code?

Thanks!

**Comment by richardliaw** [comment]
Hey @jredondopizarro,

Can you post your code?

Thanks!

**Comment by jredondopizarro** [comment]
> Hey @jredondopizarro,
> 
> Can you post your code? Or a screenshot of tensorboard? It might be possible that the ordering of returned results with Tune caused some discrepancies (due to the randomness of asynchrony). You can validate this by using `ray.init(local_mode=True)`.

I have created a jupyter notebook to illustrate my issue, you can check it out [here](https://nbviewer.jupyter.org/github/jredondopizarro/ray/blob/ray-hyperopt-loaded-trials/ray%20hp%20trials.ipynb).

**Comment by richardliaw** [comment]
@jredondopizarro this was extremely helpful.

![image](https://user-images.githubusercontent.com/4529381/62317089-c3162280-b44d-11e9-839c-f2f9c7a360c5.png)

I think this seems to work.

I made one change:

```
algo2 = HyperOptSearch(
        space,
        max_concurrent=1,
        metric="loss",
        mode = 'min',
        n_initial_points = 5,
        trials_to_load = trials_restored)
algo2.rstate.set_state(algo1.rstate.get_state())
```

Le me know if that works for you!

**Comment by richardliaw** [comment]
Also, if you're interested in contributing this, that'd be awesome! I'd be more than happy to help you get this merged.

At a high level, we would add a `HyperOptSearch.save(checkpoint_dir)` and `HyperOptSearch.restore(checkpoint_dir)`.

**Comment by jredondopizarro** [comment]
@richardliaw 
Yes, that works, thanks a lot! 
I will start working on this contribution.
